infrastructure/src/main/java/org/corfudb/infrastructure/log/LogMetadata.java

```diff
@@ -45,7 +45,11 @@ public LogMetadata() {
 
     public void update(List<LogData> entries) {
         for (LogData entry : entries) {
-            update(entry, false);
+            // This API is only used on range writes for state transfer
+            // On state transfer we need to inspect checkpoint streams (initialize = true),
+            // to avoid losing data for streams that are completely checkpointed (no actual
+            // log entry for the regular stream is present in the log).
+            update(entry, true);
```

Comment:
Can we try to remove the special handling for initialization vs post-initialization ?

Reply:
As we discussed yesterday, we can do that, but that would require changes in the stream layer to not re-sync when data is still available because it has not actually been trimmed yet. Actually maybe it's the correct way to solve it, cause if state transfer happens before the trim, we could run into this issue as well. I'll do this change accordingly.

Reply:
I was attempting to do this, but actually It won't work cause we could have data loss (if I try to optimize for these cases in the stream layer). So I believe it is better to pay the cost of a reset for state transfer scenarios only (when streams are subsumed by a checkpoint), instead of paying it for all cases (not only subsumed streams) when removing the initialize path which would enforce stream resets on regular operation (appends), state transfer and sequencer reset.